src/include/index/ivf_pq_group.h

```diff
@@ -0,0 +1,391 @@
+/**
+ * @file   ivf_pq_group.h
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2024 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * @todo This duplicates huge amounts of code from both ivf_flat and flat_pq,
+ * but there are enough differences that immediate reuse isn't obvious.  This
+ * should be refactored at some point.
+ */
+
+#ifndef TILEDB_IVF_PQ_GROUP_H
+#define TILEDB_IVF_PQ_GROUP_H
+
+#include "index/index_defs.h"
+#include "index/index_group.h"
+#include "index/ivf_pq_metadata.h"
+
+// flat_ivf_centroids, pq_ivf_centroids,
+// partitioned_pq_vectors, partitioned_pq_index, partitioned_pq_ids
+// cluster_centroids, distance_tables,
+
+[[maybe_unused]] static StorageFormat ivf_pq_storage_formats = {
+    {"0.3",
+     {
+         // @todo Should these be kept consistent with ivf_flat?
```

Comment:
I think we should pick more consistent and intuitive names for the different arrays as well as understand if all of them are actually required. I would strongly suggest to fix this before merging because changing it afterwards will lead add to have to add backwards compatibility support.

Reply:
My understanding has been that the different indexes are different and independent groups so the naming was unique to that index/group -- I just picked some names that were consistent and intuitive to me within the context of this index. But they are arbitrary so they can be anything you like, which might be useful if the same group might hold multiple types of indexes.  I can add documentation here to explain what each array is storing -- but they are all required (afaict).  `distance_tables` could perhaps be created on-demand (I have not measured the time involved, but the tables are not large).

If you would like different names for this PR, let's discuss what they should be and I can make the change(s), if any.

Loading distance tables on demand (or creating them on demand instead of storing them) is related to your previous comment.  We should probably discuss desired / expected / options for behavior and, if simple, make the changes in this PR, else push to another PR.  

(There is a similar on-demand issue with vamana regarding graph scores.)

Reply:
I would prefer to reuse the existing array names for the IVF compatible arrays. 
`pq_ivf_centroids` -> `partition_centroids`
`ivf_index` -> `partition_indexes`
`pq_ivf_vectors` -> `shuffled_vectors`
`ivf_vector_ids` -> `shuffled_vector_ids`

In theory, we could also open the array as `IVF_FLAT` and do symmetric queries if we have pq encoded query vectors.

The rest can be new names, some suggestions:
`cluster_centroids` -> `pq_cluster_centroids`
`distance_tables` -> `pq_symmetric_distance_tables`
`flat_ivf_centroids` -> `uncompressed_centroids`